component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/client/EventHubClient.java

```diff
@@ -32,74 +45,218 @@
+                            readWriteLock.writeLock().unlock();
+                            sendEvent(event);
+                        } else if (e instanceof AmqpException && isAuthenticationFailure((AmqpException) e)) {
+                            //if authentication error try to reinitialize publisher. Retrying will deal with any
+                            // network or revocation failures.
+                            log.error("Authentication issue happened. Producer client will be re-initialized "
+                                              + "retaining the EventDataBatch");
+                            this.clientStatus = ClientStatus.RETRYING;
+                            createProducerWithRetry(authEndpoint, authToken, retryOptions, false);
+                            readWriteLock.writeLock().unlock();
+                            sendEvent(event);
+                        } else if (e instanceof AmqpException && ((AmqpException) e).getErrorCondition()
+                                == AmqpErrorCondition.RESOURCE_LIMIT_EXCEEDED) {
+                            //If resource limit is exceeded we will retry after a constant delay
+                            log.error("Resource limit exceeded when publishing EventDataBatch. Operation will be "
+                                              + "retried after constant delay");
+                            try {
+                                Thread.sleep(1000 * 60);
+                            } catch (InterruptedException interruptedException) {
+                                Thread.currentThread().interrupt();
+                            }
+                            readWriteLock.writeLock().unlock();
+                            sendEvent(event);
+                        } else {
+                            //For any other exception
+                            log.error("Unknown error occurred while publishing EventDataBatch. Producer client will "
+                                              + "be re-initialized. Events may be lost in the process.");
+                            this.clientStatus = ClientStatus.RETRYING;
+                            readWriteLock.writeLock().unlock();
+                            createProducerWithRetry(authEndpoint, authToken, retryOptions, true);
+                        }
+                    } finally {
+                        try {
+                            readWriteLock.writeLock().unlock();
+                        } catch (IllegalMonitorStateException e) {
+                            //ignore
+                        }
                     }
-                } catch (InterruptedException e) {
-                    Thread.currentThread().interrupt();
-                } finally {
-                    sendSemaphore.release();
+                }
+                if (isAdded) {
+                    log.debug("Adding event: " + event.replaceAll("[\r\n]", ""));
```

Comment:
Shouldn't this be wrapped with if block to prevent string replacement happening all the time?

Reply:
Log4j2 handles this. We dont need to check isDebugEnabled

Reply:
But the arguments are evaluated when passing to the method. Since this is replacing all strings in the event, I think it shouldn't happen in each and every call.

Reply:
Fixed with 6f9169c5444fc677f7378b5e5b0d9b0a0d5fd920